frontend/dockerfile/dockerfile_addchecksum_test.go

```diff
@@ -46,6 +46,26 @@ func testAddChecksum(t *testing.T, sb integration.Sandbox) {
 		dockerfile := []byte(fmt.Sprintf(`
 FROM scratch
 ADD --checksum=%s %s /tmp/foo
+`, digest.FromBytes(resp.Content).String(), server.URL+"/foo"))
+		dir, err := integration.Tmpdir(
+			t,
+			fstest.CreateFile("Dockerfile", dockerfile, 0600),
+		)
+		require.NoError(t, err)
+		_, err = f.Solve(sb.Context(), c, client.SolveOpt{
+			LocalDirs: map[string]string{
+				builder.DefaultLocalNameDockerfile: dir,
+				builder.DefaultLocalNameContext:    dir,
+			},
+		}, nil)
+		require.NoError(t, err)
+	})
+	t.Run("DigestFromEnv", func(t *testing.T) {
+		dockerfile := []byte(fmt.Sprintf(`
+FROM scratch
+ENV DIGEST=%s
+ENV LINK=%s
+ADD --checksum=${DIGEST} ${LINK} /tmp/foo
```

Comment:
> using environment variables _and adjust them using some logic around_.

I'm a bit confused by this part, as both `ENV` and `ADD --checksum=<some_env>` are hard-coded in the Dockerfile, or was that just to avoid having to include the digest twice?


I seem to recall we added substitution for some other flags; were those flags accepting `ENV`, `ARG` (or both?)

Reply:
We should probably also update the documentation to mention that variable substitution is supported for this option (https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#verifying-a-remote-file-checksum-add---checksumchecksum-http-src-dest)

Reply:
> I'm a bit confused by this part, as both `ENV` and `ADD --checksum=<some_env>` are hard-coded in the Dockerfile, or was that just to avoid having to include the digest twice?

It is just small sample I use as test here. Having digest as an ENV will make Dockefile looks better as we will have all ENVs in one place where we put the link, the checksum (version, options, something else...). Also in case of [complex](https://github.com/lf-edge/eve/blob/365a683f23c6ea484b6f0343873c99eb23c1f395/pkg/grub/Dockerfile#L66) behavior when we use different versions (different links) for different target arches we should have control on checksum field, so we should use variables.
For example:
```
FROM foo-build-base AS foo-build-amd64
ENV FOO_VERSION=2
ENV FOO_DIGEST=sha256:...
ENV FOO_OPTIONS=...
ENV FOO_PATCHES=...

FROM foo-build-base AS foo-build-arm64
ENV FOO_VERSION=1
ENV FOO_DIGEST=sha256:...
ENV FOO_OPTIONS=...
ENV FOO_PATCHES=...

FROM foo-build-${TARGETARCH} AS foo-build
ADD --checksum=${FOO_DIGEST} ${FOO_LINK}/${FOO_VERSION}.tar.gz /foo.tar.gz
RUN build_foo...
```
Do you want me to modify the test somehow?

Reply:
> We should probably also update the documentation to mention that variable substitution is supported for this option (https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#verifying-a-remote-file-checksum-add---checksumchecksum-http-src-dest)

It seems to me that this is a separate PR with improvements in the documentation. I can see a general mention of substitution support for the ADD command here: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#environment-replacement. But going into details, we expand [source, destination](https://github.com/moby/buildkit/blob/f77133007f4b0044826169ec93c329713eb31261/frontend/dockerfile/instructions/commands.go#L257), [chown](https://github.com/moby/buildkit/blob/f77133007f4b0044826169ec93c329713eb31261/frontend/dockerfile/instructions/commands.go#L251), but they are not mentioned either. And mentioning that the checksum field supports substitution would give the wrong impression (at least to me) that other fields don't.